### PR TITLE
Lapse by votes

### PR DESCRIPTION
**Pull request by ktw1016**

### CODE REVIEW

**Comment by ktw1016** [comment]
Waiting stuff from Ross

File: client/src/charts/wrapped_nivo/wrapped_nivo_bar.js
Comment (Stephen-ONeil) [inline client/src/charts/wrapped_nivo/wrapped_nivo_bar.js:45]: Should `unsorted_initial={true}` just be the default?

**Review by Stephen-ONeil — CHANGES_REQUESTED** [review]
General feedback:
  - putting the tabbed content controls inside the height clipper feels a little clunky in use. It isn't clear they are within the collapsed section, so a first attempt to click the tab could lead to confusion when it expands the graph below it instead (which might not even happen in-view for the user). 
    - don't address this point though, I just wanted to give a warning about using the height clipper over interactive elements. Addressing the following point should make this one obsolete
  - [ ] I think the tab controls are the wrong choice for the behaviour of this panel. The expectation set up by a tab style menu is that there is different content in each tab. In this case, clicking the tab doesn't really feel like it does anything, because a) the change in the content is so minor and, in particular, b) the first piece of content under the control, the text, doesn't change at all. Could give the impression that something is broken, when the tab is first clicked (doubly so if that's actually the second click, because of the height clipper)
      - the type of control and, generally, the proximity of the control to what it effects are pretty important to consider. One "fix" for this design could be moving the text to be above the tabs so that it's more immediately clear that content changes when the tab is clicked. Better probably, because it doesn't separate the text and graph, would be to put a "toggle graph type" button somewhere (maybe in the legend, like the people graphs).
  - [ ] stats calculations and text don't cover some edge cases, take a closer look at those. E.g. [LDC](http://dev.ea-ad.ca/lapse_by_votes/index-eng.html#orgs/dept/350/infograph/financial)

File: client/src/charts/wrapped_nivo/wrapped_nivo_bar.js
Comment (ktw1016) [reply client/src/charts/wrapped_nivo/wrapped_nivo_bar.js:45]: At this point, it may as well

**Comment by ktw1016** [comment]
@Stephen-ONeil 

Actually, the text was oriented as how you described initially, above the tabbed content. But Ross specifically wanted it inside the tabbed content. So I made the change with the button and removed the tabbed content. Potential issue here is that it may not be clear to the user if they're looking at data in $ or %.

I'll have to let Ross, Val know about the change as well

**Comment by ktw1016** [comment]
>   * the type of control and, generally, the proximity of the control to what it effects are pretty important to consider. One "fix" for this design could be moving the text to be above the tabs so that it's more immediately clear that content changes when the tab is clicked. Better probably, because it doesn't separate the text and graph, would be to put a "toggle graph type" button somewhere (maybe in the legend, like the people graphs).

Actually, the text was oriented as how you described initially, above the tabbed content. But Ross specifically wanted it inside the tabbed content. So I made the change with the button and removed the tabbed content. Potential issue here is that it may not be clear to the user if they're looking at data in $ or %.

**Comment by rossbm** [comment]
> > * I think the tab controls are the wrong choice for the behaviour of this panel. The expectation set up by a tab style menu is that there is different content in each tab. In this case, clicking the tab doesn't really feel like it does anything, because a) the change in the content is so minor and, in particular, b) the first piece of content under the control, the text, doesn't change at all. Could give the impression that something is broken, when the tab is first clicked (doubly so if that's actually the second click, because of the height clipper)
> > * the type of control and, generally, the proximity of the control to what it effects are pretty important to consider. One "fix" for this design could be moving the text to be above the tabs so that it's more immediately clear that content changes when the tab is clicked. Better probably, because it doesn't separate the text and graph, would be to put a "toggle graph type" button somewhere (maybe in the legend, like the people graphs).
> 
> Actually, the text was oriented as how you described initially, above the tabbed content. But Ross specifically wanted it inside the tabbed content. So as of now, I made the change with the button to toggle between $ and %. Also removed the tabbed content.
> 
> Potential issue here is that it may not be clear to the user if they're looking at data in $ or %.

The button on the graph looks good. Maybe just capitalize every word in text? So "Show Lapsed Authorities in $/%".

**Comment by rossbm** [comment]
> stats calculations and text don't cover some edge cases, take a closer look at those. E.g. LDC

Good catch!

I guess to catch LDC and similar, years with $0 (or missing) authorities should be excluded when calculating the average lapse at the department level. Can't think of any other edge cases off the top of my head.

**Comment by ktw1016** [comment]
Screen Shot 2021-01-06 at 6 33 00 PM

It would look like that with it capitalized. Is that good?

**Comment by ktw1016** [comment]
Hi @rossbm , Val had a look, she wanted to emphasize the fact that lapsing funds is normal and expected operations in government (basically what 2nd paragraph of our lapse glossary). She also seemed opposed to capitalizing all the words because some people may feel it's kinda screaming at them

`Lapsing funds is a normal and expected part of any budgetary process` I think this should be sufficient, no need for extra translations because it's already in the glossary.

What do you think? Let me know if you had anything in mind to add on to that!

**Comment by rossbm** [comment]
> Hi @rossbm , Val had a look, she wanted to emphasize the fact that lapsing funds is normal and expected operations in government (basically what 2nd paragraph of our lapse glossary). She also seemed opposed to capitalizing all the words because some people may feel it's kinda screaming at them
> 
> `Lapsing funds is a normal and expected part of any budgetary process` I think this should be sufficient, no need for extra translations because it's already in the glossary.

Did she want extra text added to accompany the chart? That sentence can work at the end of the paragraph that accompanies that lapse chart. Or maybe it should go accompany the first chart in the panel, since that panel also talks about unspent funds. Best to provide all necessary explanations upfront.


For the toggling between % and $, can we make the it into a radio button? Or maybe some sort of slider? Not certain how easy that would be. The radio button would show that there are two ways of viewing the chart.

I suggested capitalizing the words in "Show Lapsed Authorities in $/%" mainly we capitalize "Toggle Graph Type" in the People data. I never know when things should be capitalized or not.

**Comment by ktw1016** [comment]
@rossbm I think Val essentially wanted to be very clear that lapsing is normal and expected operations. She did mention to put it below the current text, but I suppose we could also put it up with auth/exp. I don't see much differences in that though.

For the radio button or slider, it should be easy to do. All required is to change up the look of it. I'll wait to hear what @Stephen-ONeil  thinks of it!

**Comment by Stephen-ONeil** [comment]
A `RadioButtons` sounds good to me!

**Comment by rossbm** [comment]
> @rossbm I think Val essentially wanted to be very clear that lapsing is normal and expected operations. She did mention to put it below the current text, but I suppose we could also put it up with auth/exp. I don't see much differences in that though.

@ktw1016 I think that just putting the text below the current text is best and easiest for now. Ideally we would make the similarities and differences between the two charts a bit clearer, but that's a problem for another time 😊

**Review by Stephen-ONeil — CHANGES_REQUESTED** [review]
Oh, I was doing a review but I see you're still making commits. I'll just leave these couple comments for now and re-review when you're done.

**Comment by ktw1016** [comment]
@Stephen-ONeil I'm done now, I was just updating with the radio buttons

File: client/src/charts/legends/StandardLegend.js
Comment (ktw1016) [reply client/src/charts/legends/StandardLegend.js:45]: So this was when I had the toggle button inside the legend. I got rid of the button inside legend in favor of a different location, and now a radio button. But just left this change in case we want multiple controls in the legend. Then it'd be better for aesthetics to have dashed border between each control.

I could remove this change if you don't like it

File: client/src/panels/panel_declarations/finances/auth_exp_planned_spending/auth_exp_planned_spending.js
Comment (Stephen-ONeil) [inline client/src/panels/panel_declarations/finances/auth_exp_planned_spending/auth_exp_planned_spending.js:287]: I find `_.includes([1,2,3], votestattype)` to be a good shorthand for this kind of check.

File: client/src/panels/panel_declarations/finances/auth_exp_planned_spending/auth_exp_planned_spending.js
Comment (Stephen-ONeil) [inline client/src/panels/panel_declarations/finances/auth_exp_planned_spending/auth_exp_planned_spending.js:372]: Did you consider the `RadioButtons` component? Might look a little fancier, up to you though (let me know if you preferred standard radio buttons for some reason).

**Review by Stephen-ONeil — CHANGES_REQUESTED** [review]
Haha, the big thing I was going to suggest changing was the centred button stuff that your last commit changed, so all I've got are more little comments!

File: client/src/components/DisplayTable/DisplayTable.js
Comment (ktw1016) [reply client/src/components/DisplayTable/DisplayTable.js:45]: hahaha did you first come up with that name?

Comment (Stephen-ONeil) [reply client/src/components/DisplayTable/DisplayTable.js:45]: Don't think so, maybe Kate? I think it was always kind of a rough-draft of a name, we just never gave it a better one.

File: client/src/panels/panel_declarations/finances/auth_exp_planned_spending/auth_exp_planned_spending.js
Comment (ktw1016) [reply client/src/panels/panel_declarations/finances/auth_exp_planned_spending/auth_exp_planned_spending.js]: Didn't know we had RadioButtons component😂, Changed it, I think it looks better but kinda resemble the tabbed control. Oh well, I'm sticking with it.

File: client/src/components/DisplayTable/DisplayTable.js
Comment (ktw1016) [reply client/src/components/DisplayTable/DisplayTable.js]: Couldn't think of another name, sticking with DisplayTable haha

File: data/test-data/org_vote_stat_pa.csv
Comment (Stephen-ONeil) [inline data/test-data/org_vote_stat_pa.csv:2]: CBSA isn't in the test IGOC. Think you wanted CPCC. The snapshot tests specifically queries for CPCC data, so losing it is kind of undermining the snapshot outputs worth.

Side note, I'd recommend making minimal changes whenever updating the test data. We want it to be clear from the snapshot diff that any implementation changes only resulted in the expected output changes. Any time the whole input data/output snapshot changes all at once, it's kind of a loss of confidence in the whole snapshot test itself (unless you then examine the whole snapshot output and manually confirm it).

Fine here because we saw in that earlier snapshot diff that the other fields hadn't changed, just a general caution on snapshot tests.

File: client/src/charts/legends/StandardLegend.js
Comment (Stephen-ONeil) [reply client/src/charts/legends/StandardLegend.js:45]: Sorry, that thumbs up was supposed to mean "yes, please remove this change, thanks". Should have just clearly responded, haha.

**Review by Stephen-ONeil — CHANGES_REQUESTED** [review]

File: client/src/models/populate_survey.js
Comment (Stephen-ONeil) [inline client/src/models/populate_survey.js:27]: Is this file from a different branch?

Comment (ktw1016) [reply client/src/models/populate_survey.js:27]: Oops, this file wasn't supposed to be here haha

**Review by Stephen-ONeil — APPROVED** [review]
Seems all good code-wise, don't know if there's content review/approval left before merging but otherwise I approve.

Kind of hard to highlight the new panel with a featured content link, etc., since the new content is under old content and starts collapsed. Hmm...

**Comment by ktw1016** [comment]
I don't think Val want to highlight this on featured content or anything, because lapse can potentially be sensitive(?) at times. If I remember correctly, she liked having it collapsed in the height clipper haha

**Comment by ktw1016** [comment]
This is just waiting for Glenn's approval

**Review by Stephen-ONeil — CHANGES_REQUESTED** [review]
Was taking a quick look, colours on the line segments are broken.

image

I'll look a bit more. Please also take a quick review of this branch, @ktw1016, so we can merge it in (ideally today or Monday).

**Comment by ktw1016** [comment]
@Stephen-ONeil fixed! I'm taking a look at it, not much jumping out to me

**Review by Stephen-ONeil — CHANGES_REQUESTED** [review]
Huh, how'd `client/src/panels/panel_declarations/finances/covid/covid_estimates.js` get back in here? Can delete that file, the current version is in `panel_declarations/covid`, not under `/finances`.

Please confirm that the changes in the .csv files are right and will be on the pipeline's master branch when/right after this merges!

**Pull request merged**
